README.md
# Data sets

Traffic fatality data sets for Austin, TX.

## Data availability

The traffic fatality reports pusblished by APD are hosted on their website for 2 years. Past this point they are
archived and are not publicly accessible anymore.

## Workflow

Our automated workflow is composed of the 4 following steps:

1. Generate the raw data sets.
2. Import/Merge the external data sets.
3. Augment the data sets.
4. Merge the results into `*-all-*.json` data sets.

Using these techniques to generate the data sets, we ensure that they can always be recreated in case of problem would occur.

Each step will be detailed in a dedicated section bellow.

### Generate raw data sets

The data sets named `fatalities-{year}-raw.json` are generated directly from `ScrAPD` without any manual intervention. A data set is created for each year.

### Import/Merge external data sets

Each of these external data sets has a dedicated documentation file in the `docs` folder.

* Socrata

Here is an example showing how the data is imported:

```bash
TOPDIR=$(git rev-parse --show-toplevel); for YEAR in {17..18}; do python ${TOPDIR}/tools/scrapd-importer-fatalities-socrata.py ${TOPDIR}/datasets/fatalities-20${YEAR}-raw.json ${TOPDIR}/external-datasets/socrata-apd-archives/socrata-apd-20${YEAR}.json > ${TOPDIR}/datasets/fatalities-20${YEAR}-augmented.json;done
```

### Augment the data sets

The data sets named `fatalities-{year}-augmented.json` are data sets that have been enhanced, in order to improve the
quality of the data.

```bash
# Generate empty data sets.
for f in fatalities-20{13..19}-augmented.json; do echo "[]" > "${f}"; done
```

Augment the data sets:

```bash
for f in fatalities-20{13..19}-augmented.json; do python "${TOPDIR}/tools/scrapd-augmenter-geocoding-geocensus.py"-i ${f}; done
```

The data sets are being augmented with ScrAPD augmenters that can be found in the `tools` folder. Each augmenter

Currently the following augmenters are available:

* scrapd-augmenter-geocoding-geocensus

#### Manual corrections

Corrections can also be added manually. Corrections can add extra fields or update values in existing fields. They are
applied last.

A correction must be made in a file named `fatalities-{year}-manual.json`. All the corrections for the same year
**MUST** be grouped together in the same file. The order does not matter. If an entry is found several times, the last
one (i.e. the lowest one in the file) will superseed all the others.

```json
[
  {
    "19-0400694": {
      "Type": "Pedestrian",
      "Gender": "Female",
    },
    "19-0320079": {
      "Type": "Bicyle",
      "Gender": "Unknown",
    }
  }
]
```

### "all" data sets

The data sets whose year is `all` are a combination of all the data sets of the same category.

There are generated using the following `jq` command:

```bash
jq -s add fatalities-20{17..19}-raw.json > fatalities-all-raw.json
jq -s add fatalities-20{17..19}-augmented.json > fatalities-all-raw.json
```

## Full updates

A full update happens when ScrAPD gets updated with changes that drastically improve the quality of the data which was retrieved.

ScrAPD versions which required a full update:

* 1.5.0
* 1.5.1

As a result, all the data sets were updated with the following command:

```bash
for i in {17..19}; do python tools/scrapd-merger.py -i fatalities-20${i}-raw.json <(scrapd -v --format json --from "Jan 1 20${i}" --to "Dec 31 20${i}"); done
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The data sets being potentially updated daily, the Changelog does not contain information regarding the updates of the
data, unless they are of crucial importance.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Calendar Versioning](https://calver.org/).

## [Unreleased]

## 19.7.17.0

- Regenerate raw data sets with ScrAPD 2.1.0.

## 19.6.8.0

## Changed

- Change the Socrata augmentations to include the "Speeding" field.

## 19.4.8.0

### Added

- Added an augmented data set including the geolocation of the crashes.

## 19.3.23.0

### Fixed

- Manually normalized the date fields of the entries retrieved prior to ScrAPD 1.5.0.

## 19.3.19.0

### Added

- Added a tool to merge two ScrAPD data sets together.

### Changed

- Changed the calver format from `{yy}.{month}.{PATCH}` to `{yy}.{month}.{dom}.{PATCH}`
- Use ScrAPD `1.5.0` to update the datasets.

## 19.3.0

Initial release.

[//]: # (Release links)

[//]: # (Issue/PR links)



tools/update-datasets.sh
#!/bin/bash
set -euo pipefail

# Define variables.
: ${REGENERATE:=0}
CURRENT_YEAR=$(date +%Y)
TOPDIR=$(git rev-parse --show-toplevel)
DATASET_DIR="${TOPDIR}/datasets"
CURRENT_DATASET="${DATASET_DIR}/fatalities-${CURRENT_YEAR}-raw.json"
TOOL_DIR="${TOPDIR}/tools"
MERGER="${TOPDIR}/tools/scrapd-merger.py"

# Ensure we are in the top directory.
cd "${TOPDIR}"|| exit

# Generate the current data set.
ENTRY_COUNT_BEFORE=$(jq length "${CURRENT_DATASET}")
[ "$REGENERATE" == 0 ] && python "${MERGER}" -i "${CURRENT_DATASET}" <(scrapd -vv --format json --from "Jan 1 ${CURRENT_YEAR}" --to "Dec 31 ${CURRENT_YEAR}");
HAS_CHANGE=$(git status -s)

# If nothing changed, we can leave.
if [ -z "${HAS_CHANGE}" ]; then
  echo "There is nothing new to commit."
  [ "$REGENERATE" == 0 ] && exit 0
fi

for YEAR in {2017..2019}; do
  echo "=> Processing year ${YEAR}..."

  # Prepare variables.
  RAW_DATA_SET="${DATASET_DIR}/fatalities-${YEAR}-raw.json"
  AUGMENTED_DATA_SET="${DATASET_DIR}/fatalities-${YEAR}-augmented.json"
  AUGMENTATION_DIR="${TOPDIR}/augmentations/${YEAR}"

  # Create augmented data sets from the raw ones.
  echo -e "\t- Resetting augmented data sets..."
  cp "${RAW_DATA_SET}" "${AUGMENTED_DATA_SET}"

  # Apply the augmentations (1st pass).
  # This is to restore the previous state as we rebuilt the data set from scratch.
  echo -e "\t- Applying augmentations (1st pass)..."
  for AUGMENTATION in "${AUGMENTATION_DIR}"/*.json; do
    echo -e "\t\t- $(basename ${AUGMENTATION})"
    python "${MERGER}" -i "${AUGMENTED_DATA_SET}" "${AUGMENTATION}"
  done

  # Generate the augmentations.
  echo -e "\t- Generating new augmentations..."
  AUGMENTATIONS="scrapd-augmenter-geocoding-geocensus.py:augmentation-geocoding-geocensus-${YEAR}.json"
  for AUGMENTATION in ${AUGMENTATIONS}; do
    TOOL="${TOOL_DIR}/$(echo ${AUGMENTATION}|cut -d':' -f1)"
    AUGMENTATION_FILE="${AUGMENTATION_DIR}/$(echo ${AUGMENTATION}|cut -d':' -f2)"
    echo -e "\t\t- $(basename ${AUGMENTATION_FILE})"
    python "${TOOL}" "${AUGMENTED_DATA_SET}" > "${AUGMENTATION_FILE}"
  done

  # Apply the augmentations (2nd pass).
  # This is to add the new augmentations if any.
  echo -e "\t- Applying augmentations (2nd pass)..."
  for AUGMENTATION in "${AUGMENTATION_DIR}"/*.json; do
    echo -e "\t\t- $(basename ${AUGMENTATION})"
    python "${MERGER}" -i "${AUGMENTED_DATA_SET}" "${AUGMENTATION}"
  done
done

# Merge the results.
cd "${TOPDIR}/datasets"|| exit
echo "=> Merging the yearly data sets..."
jq -s add fatalities-20{17..19}-raw.json > fatalities-all-raw.json
jq -s add fatalities-20{17..19}-augmented.json > fatalities-all-augmented.json

# Compute the number of new entries.
ENTRY_COUNT_AFTER=$(jq length "${CURRENT_DATASET}")
NEW_ENTRY_COUNT=$(( ENTRY_COUNT_AFTER - ENTRY_COUNT_BEFORE ))
echo "There are ${NEW_ENTRY_COUNT} new entries in the current data set."
[ "$REGENERATE" == 1 ] && exit

# Go back to the top dir.
cd "${TOPDIR}"|| exit

# Commit the changes.
git add .
git commit -m "Update data sets" \
  -m "There are ${NEW_ENTRY_COUNT} new entries in the current data set." \
  -m "$(git status -s)"
pycalver bump -n || pycalver bump -n --patch
git push origin master
git push --tags
